utils/scarb-ui/src/lib.rs

```diff
@@ -149,7 +179,24 @@ impl Ui {
 
     fn do_print<T: Message>(&self, message: T) {
         match self.output_format {
-            OutputFormat::Text => message.print_text(),
+            OutputFormat::Text => {
+                let read_lock = self.state.read().expect("failed to acquire read lock");
+                if let Some(widget) = read_lock.active_print_overrider.as_ref() {
+                    if let Some(handle) = widget.upgrade() {
+                        handle.println(message.text().as_str());
+                    } else {
+                        message.print_text();
+                        // Clear widget from state.
+                        drop(read_lock);
+                        self.state
+                            .write()
+                            .expect("failed to acquire write lock")
+                            .active_print_overrider = None;
+                    }
+                } else {
+                    message.print_text()
+                }
```

Comment:
I sense several issues here:
1. This seems to hold the read lock for too long.
2. This write seems unnecessary? If widget dies, then the weak will always fail to upgrade so this looks to be effectively no-op

FYI Take a look at this: https://gcc.godbolt.org/z/rfj7MEG3s. It looks like, in hot path and aggressive optimizations, `Weak::upgrade` is almost the same cost-wise (here is even better -- fewer instructions) as matching over `Option`.

Heck, you can even replace `Option<Weak<...>>` with just `Weak` and use `Weak::new()` as `None` replacement.